// filestack-js/build/main/lib/filelink.spec.js
"use strict";
/*
 * Copyright (c) 2019 by Filestack.
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var filelink_1 = require("./filelink");
var transforms_schema_1 = require("./../schema/transforms.schema");
var validator = require("../schema");
describe('filelink', function () {
    beforeAll(function () {
        spyOn(validator, 'getValidator').and.callThrough();
    });
    var defaultSource = '5aYkEQJSQCmYShsoCnZN';
    var defaultApikey = 'DEFAULT_API_KEY';
    it('should properly instantiate Filelink', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        expect(filelink).toBeDefined();
        expect(filelink).toBeInstanceOf(filelink_1.Filelink);
    });
    it('should throw an error when handle is invalid', function () {
        var source = '*/5aYkEQJSQCmYShsoCnZN';
        expect(function () {
            var f = new filelink_1.Filelink(source);
        }).toThrow('Invalid filestack source provided');
    });
    it('should throw an error when external handle and without apikey', function () {
        var source = 'src://test123/example.jpg';
        expect(function () {
            var filelink = new filelink_1.Filelink(source);
        }).toThrow('External sources requires apikey to handle transforms');
    });
    it('should be able to convert filelink to string', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        var result = filelink.toString();
        expect(result).toBe('https://cdn.filestackcontent.com/5aYkEQJSQCmYShsoCnZN');
    });
    it('should allow to set validator status with toString method', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        filelink.setUseValidator(false);
        expect(filelink.resize({
            // @ts-ignore
            t: 1,
        }).toString()).toEqual('https://cdn.filestackcontent.com/resize=t:1/5aYkEQJSQCmYShsoCnZN');
    });
    it('should allow to set validator status with getTransformations method', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        filelink.setUseValidator(false);
        filelink.resize({
            // @ts-ignore
            t: 1,
        });
        expect(filelink.getTransformations()).toEqual([{
                name: 'resize',
                params: { t: 1 },
            }]);
    });
    it('should create filelink with apikey when is provided', function () {
        var filelink = new filelink_1.Filelink(defaultSource, defaultApikey);
        var result = filelink.toString();
        expect(result).toBe('https://cdn.filestackcontent.com/DEFAULT_API_KEY/5aYkEQJSQCmYShsoCnZN');
    });
    it('should be able to use many tasks at once and reset them', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        var resizeParams = {
            width: 200,
        };
        var rotateParams = {
            deg: 90,
        };
        filelink
            .resize(resizeParams)
            .rotate(rotateParams)
            .crop({
            dim: [20, 20, 250, 250],
        });
        expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/resize=width:200/rotate=deg:90/crop=dim:[20,20,250,250]/5aYkEQJSQCmYShsoCnZN');
        filelink.reset();
        expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/5aYkEQJSQCmYShsoCnZN');
    });
    it('should be able to get transformations', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        var resizeParams = {
            width: 200,
        };
        var rotateParams = {
            deg: 90,
        };
        filelink
            .resize(resizeParams)
            .rotate(rotateParams)
            .crop({
            dim: [20, 20, 250, 250],
        });
        var expected = [{ name: 'resize', params: { width: 200 } }, { name: 'rotate', params: { deg: 90 } }, { name: 'crop', params: { dim: [20, 20, 250, 250] } }];
        expect(filelink.getTransformations()).toEqual(expected);
    });
    it('should be able to getValidationSchema', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        var result = filelink.getValidationSchema();
        expect(result).toEqual(transforms_schema_1.TransformSchema);
    });
    it('should not require apikay on filestack external url', function () {
        var filelink = new filelink_1.Filelink('https://cdn.filestackcontent.com/5aYkEQJSQCmYShsoCnZN');
        filelink.shadow(false).upscale();
        expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/upscale/\"https://cdn.filestackcontent.com/5aYkEQJSQCmYShsoCnZN\"');
    });
    it('should be able to disable selected task', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        filelink.shadow(false).upscale();
        expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/upscale/5aYkEQJSQCmYShsoCnZN');
    });
    it('should be able to create filelink for many handles', function () {
        var sourceArr = ['5aYkEQJSQCmYShsoCnZN', '4aYkEQJSQCmYShsoCnZN'];
        var filelink = new filelink_1.Filelink(sourceArr);
        expect(filelink.toString()).toEqual('https://cdn.filestackcontent.com/[5aYkEQJSQCmYShsoCnZN,4aYkEQJSQCmYShsoCnZN]');
        filelink.setBase64(true);
        expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/b64://WzVhWWtFUUpTUUNtWVNoc29DblpOLDRhWWtFUUpTUUNtWVNoc29DblpOXQ==');
    });
    it('should be able to create filelink for many src handles', function () {
        var sourceArr = ['src://test123/example.jpg', 'src://test123/flug_9-trans_atlantik-400dpi.jpg'];
        var filelink = new filelink_1.Filelink(sourceArr, defaultApikey);
        expect(filelink.toString()).toEqual('https://cdn.filestackcontent.com/DEFAULT_API_KEY/["src://test123/example.jpg","src://test123/flug_9-trans_atlantik-400dpi.jpg"]');
    });
    it('should throw an error if task params are not valid', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        expect(function () {
            // @ts-ignore
            filelink.resize({ r: 2 }).toString();
        }).toThrow('Params validation error');
    });
    it('should omit transformations with empty options', function () {
        var filelink = new filelink_1.Filelink(defaultSource);
        filelink.upscale(false);
        expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/5aYkEQJSQCmYShsoCnZN');
    });
    it('should throw an error if source does not exists', function () {
        expect(function () {
            return new filelink_1.Filelink('');
        }).toThrow('Source not Set');
    });
    describe('Different tasks', function () {
        var filelink = new filelink_1.Filelink(defaultSource, defaultApikey);
        afterEach(function () {
            filelink = new filelink_1.Filelink(defaultSource, defaultApikey);
        });
        it('should be able to create filelink when handle is base64', function () {
            filelink.upscale();
            filelink.setBase64(true);
            expect(filelink.toString()).toBe('https://cdn.filestackcontent.com/DEFAULT_API_KEY/b64/W3sibmFtZSI6InVwc2NhbGUifV0=/b64://NWFZa0VRSlNRQ21ZU2hzb0NuWk4=');
        });
        it('should be able to use custom cname', function () {
            var cname = 'newcname.com';
            filelink.setCname(cname);
            expect(filelink.toString()).toBe('https://cdn.newcname.com/DEFAULT_API_KEY/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to use custom domain', function () {
            var customDomain = 'https://customDomain.com';
            filelink.setCustomDomain(customDomain);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to autoImage transformation', function () {
            filelink.autoImage();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/auto_image/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to add flip', function () {
            filelink.flip();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/flip/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to add flop', function () {
            filelink.flop();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/flop/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to enhance', function () {
            filelink.enhance();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/enhance/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to redeye', function () {
            filelink.redeye();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/redeye/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to monochrome', function () {
            filelink.monochrome();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/monochrome/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to negative', function () {
            filelink.negative();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/negative/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to tags', function () {
            filelink.tags();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/tags/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to sfw', function () {
            filelink.sfw();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/sfw/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to store', function () {
            var storeParams = {};
            filelink.store(storeParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/store/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to disable cache', function () {
            filelink.cache(false);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/cache=false/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to set cache params', function () {
            var cacheParams = {
                expiry: 666,
            };
            filelink.cache(cacheParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/cache=expiry:666/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able add animation transform', function () {
            var params = {
                loop: 1,
                delay: 10,
                width: 10,
                height: 10,
                fit: filelink_1.FitOptions.scale,
                background: 'transparent',
                align: [filelink_1.Align.top, filelink_1.Align.center],
            };
            filelink.animate(params);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/animate=loop:1,delay:10,width:10,height:10,fit:scale,background:transparent,align:[top,center]/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to resize', function () {
            var resizeParams = {
                width: 200,
            };
            filelink.resize(resizeParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/resize=width:200/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to crop', function () {
            filelink.crop({
                dim: [20, 20, 250, 250],
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/crop=dim:[20,20,250,250]/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to rotate', function () {
            var rotateParams = {
                deg: 90,
            };
            filelink.rotate(rotateParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/rotate=deg:90/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to detect faces', function () {
            filelink.detectFaces({});
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/detect_faces/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to crop faces', function () {
            filelink.cropFaces({});
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/crop_faces/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to pixelate faces', function () {
            filelink.pixelateFaces({});
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/pixelate_faces/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to blur faces', function () {
            var blurFacesParams = {
                amount: 3,
                blur: 5,
                type: filelink_1.ShapeType.oval,
            };
            filelink.blurFaces(blurFacesParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/blur_faces=amount:3,blur:5,type:oval/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to round cornerns', function () {
            var roundedCornersParams = {
                radius: 50,
                blur: 5,
            };
            filelink.roundedCorners(roundedCornersParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/rounded_corners=radius:50,blur:5/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to polaroid', function () {
            var polaroidParams = {
                color: 'ff0000',
                rotate: 50,
            };
            filelink.polaroid(polaroidParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/polaroid=color:ff0000,rotate:50/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to vignette', function () {
            var vignetteParams = {
                background: 'ff0000',
                amount: 5,
            };
            filelink.vignette(vignetteParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/vignette=background:ff0000,amount:5/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to tornEdges', function () {
            filelink.tornEdges({});
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/torn_edges/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to shadow', function () {
            filelink.shadow();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/shadow/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to circle', function () {
            filelink.circle({});
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/circle/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to border', function () {
            filelink.border();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/border/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to vignette', function () {
            filelink.vignette({});
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/vignette/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to sharpen', function () {
            var sharpenParams = {
                amount: 5,
            };
            filelink.sharpen(sharpenParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/sharpen=amount:5/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to blur', function () {
            var blurParams = {
                amount: 5,
            };
            filelink.blur(blurParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/blur=amount:5/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to blackwhite', function () {
            var blackwhiteParams = {
                threshold: 50,
            };
            filelink.blackwhite(blackwhiteParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/blackwhite=threshold:50/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to sepia', function () {
            filelink.sepia();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/sepia/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to pixelate', function () {
            filelink.pixelate();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/pixelate/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to oilPaint', function () {
            filelink.oilPaint();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/oil_paint/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to modulate', function () {
            filelink.modulate();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/modulate/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to partialPixelate', function () {
            filelink.partialPixelate({
                objects: [[20, 20, 50, 50]],
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/partial_pixelate=objects:[[20,20,50,50]]/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to partialBlur', function () {
            filelink.partialBlur({
                amount: 5,
                objects: [[20, 20, 50, 50]],
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/partial_blur=amount:5,objects:[[20,20,50,50]]/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to collage', function () {
            filelink.collage({
                files: ['http://welcome-swiss.com/wp-content/uploads/2015/12/Swiss-landscape.jpg'],
                width: 200,
                height: 200,
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/collage=files:["http://welcome-swiss.com/wp-content/uploads/2015/12/Swiss-landscape.jpg"],width:200,height:200/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to upscale', function () {
            filelink.upscale();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/upscale/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to ascii', function () {
            filelink.ascii();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/ascii/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to quality', function () {
            var qualityParams = {
                value: 5,
            };
            filelink.quality(qualityParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/quality=value:5/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to use security', function () {
            var securityParams = {
                policy: 'examplePolicy',
                signature: 'exampleSignature',
            };
            filelink.security(securityParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/security=policy:examplePolicy,signature:exampleSignature/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able output', function () {
            var outputParams = {
                format: 'png',
            };
            filelink.output(outputParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/output=format:png/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to videoConvert', function () {
            var videoConvertParams = {
                aspect_mode: filelink_1.VideoAccessMode.letterbox,
                upscale: true,
            };
            filelink.videoConvert(videoConvertParams);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/video_convert=aspect_mode:letterbox,upscale:true/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to URLScreenshot', function () {
            filelink.URLScreenshot();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/urlscreenshot/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to PDFInfo', function () {
            filelink.PDFInfo();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/pdfinfo/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to PDFConvert', function () {
            filelink.PDFConvert();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/pdfconvert/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to force cache', function () {
            filelink.cache(true);
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/cache/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to add fallback transformation', function () {
            filelink.fallback({
                handle: 'http:test.com',
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/fallback=handle:http:test.com/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to add zip trnasformation', function () {
            filelink.zip();
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/zip/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to add minify js transformation', function () {
            filelink.minifyJs({
                mangle: true,
                gzip: true,
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/minify_js=mangle:true,gzip:true/5aYkEQJSQCmYShsoCnZN');
        });
        it('should be able to add minify css transformation', function () {
            filelink.minifyCss({
                level: 1,
                gzip: true,
            });
            expect(filelink.toString()).toBe('https://customDomain.com/DEFAULT_API_KEY/minify_css=level:1,gzip:true/5aYkEQJSQCmYShsoCnZN');
        });
    });
});
